### PR TITLE
Ux love

### PR DESCRIPTION
**Pull request by lukasbuenger**
Depends on orbiting/styleguide#81

Changelog:

Many small changes, but most notably it is possible to define static fields through editorOptions. Also there is a way to formulate a "definition of empty" on composed components that will get checked when Delete or Backspace is pressed. 

Figure Byline with inline placeholders

Sidebar open by default and click on cogwheel prevents the editor from losing focus.

### CODE REVIEW

File: components/editor/utils/keyHandlers.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    if (isTab) {
+      if (nextBlock) {
+        event.preventDefault()
+        return focusNext(change)
+      }
+      return
+    }
+
+    if (isEnter) {
+      event.preventDefault()
```
Comment (tpreusse) [inline components/editor/utils/keyHandlers.js]: missing support for soft break with shift: `change.insertText('\n')`

since the lead becomes a static one this is mandatory—we need soft break support in leads

Comment (lukasbuenger) [reply components/editor/utils/keyHandlers.js]: This will have to be exposed in the template, I guess. While soft breaks in leads are ok, in figure captions et al they're certainly not.
Will do.

Comment (tpreusse) [reply components/editor/utils/keyHandlers.js]: already supported now and exposed in template. why not in figure captions?

Comment (lukasbuenger) [reply components/editor/utils/keyHandlers.js]: Because it looks ugly? Where is this exposed in template?
I mean, we could turn that for all static fields by default, no problem.

File: components/editor/modules/figuregroup/ui.js [hunk 2]
```diff
@@ -127,24 +127,33 @@ export const FigureGroupForm = options => {
...
+
+  const insertTypes = options.rule.editorOptions.insertTypes || []
   return ({ value, onChange }) => {
-    const disabled = value.isBlurred
+    const disabled = value.isBlurred ||
+      !value.blocks.every(
+        n => insertTypes.includes(n.type)
+      )
+
+       // matchBlock(options.TYPE), value).size > 0
```
Comment (uxengine) [inline components/editor/modules/figuregroup/ui.js]: remove?

File: components/editor/utils/keyHandlers.js (hunk 1 above)
Comment (tpreusse) [reply components/editor/utils/keyHandlers.js]: yes please turn on by default to match status quo. Br is part of the `globalInlines` in the template.

File: components/editor/Placeholder.js [hunk 3]
```diff
@@ -17,3 +18,19 @@ export default ({children}) => {
     </span>
   )
 }
+
+const styles = {
+  inline: css({
+    pointerEvents: 'none',
+    opacity: '0.333',
```
Comment (uxengine) [inline components/editor/Placeholder.js]: don't think this number needs quotes, just `.333` should be enough.

File: components/editor/utils/keyHandlers.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    const isBackspace = event.key === 'Backspace'
+    const isEnter = event.key === 'Enter'
+    const isTab = event.key === 'Tab'
+    const isDelete = event.key === 'Delete'
+
+    if (!isBackspace && !isEnter && !isDelete && !isTab) return
+    const { value } = change
+    const inSelection = value.blocks.some(matchBlock(TYPE))
+
+    if (inSelection && value.isExpanded) {
```
Comment (tpreusse) [inline components/editor/utils/keyHandlers.js]: this feels to invasive: static one can now only be delete on character at a time. Could we support deleting a range if it is exclusively within one/our block? And when not first reduce the range to just that block? so double backspace would always work to clear a complete lead / credit line

File: components/editor/utils/keyHandlers.js (hunk 1 above)
Comment (lukasbuenger) [reply components/editor/utils/keyHandlers.js]: Will do.
You mean its supported through the presence of a certain serializer. But I really wouldn't want to loop through the child serializers to determine whether the UI should capture soft breaks or not.

File: components/editor/modules/figure/caption.js [hunk 5]
```diff
@@ -1,12 +1,14 @@
 import MarkdownSerializer from 'slate-mdast-serializer'
 import { Block } from 'slate'
 import { matchBlock } from '../../utils'
-import Placeholder from '../../Placeholder'
+import { staticKeyHandler } from '../../utils/keyHandlers'
+import { Inline } from '../../Placeholder'
 
 const getSerializer = options => {
+  const [ byLineModule, ...subModules ] = options.subModules
```
Comment (uxengine) [inline components/editor/modules/figure/caption.js]: `byline` is one word and doesn't necessarily need to be camelCased `byLine`, but I leave that up to you:
https://en.wikipedia.org/wiki/Byline

**Comment by tpreusse** [comment]
Agree. No need to loop through—just do it globally. All module that start to use `staticKeyHandler` had the support for soft break before expect for `figureCaption` but that was an oversight when reviewing. I agree that it shouldn't be used often in figure captions but there are edge cases that it can be useful for.

File: components/editor/utils/keyHandlers.js (hunk 1 above)
Comment (tpreusse) [reply components/editor/utils/keyHandlers.js]: Agree. No need to loop through—just do it globally. All module that start to use `staticKeyHandler` had the support for soft break before expect for `figureCaption` but that was an oversight when reviewing. I agree that it shouldn't be used often in figure captions but there are edge cases that it can be useful for.

**Comment by tpreusse** [comment]
The test are current broken, with locally linked styleguide:
```
15:13 ~/Code/publikator-frontend [ux-love]$ npm test

> publikator-frontend@1.0.0 test /Users/thomas/Code/publikator-frontend
> cross-env NODE_ENV=test babel-tape-runner "?(lib|pages|components)/**/*.test.js" | faucet

/Users/thomas/Code/publikator-frontend/components/editor/modules/figure/caption.js:98
  } = byLineModule.rule.editorOptions || {};
                        ^

TypeError: Cannot read property 'editorOptions' of undefined
```

**Comment by lukasbuenger** [comment]
Sure. Without the new styleguide template a lot of stuff doesn't work.

**Comment by tpreusse** [comment]
@lukasbuenger I test with your styleguide branch ;)

**Comment by tpreusse** [comment]
Overall this is a great improvements. Once soft-break and range delete in static is addressed I'm happy to approve it.

Thank you @lukasbuenger

File: components/editor/modules/figure/caption.js [hunk 6]
```diff
@@ -35,16 +37,31 @@ const getSerializer = options => {
     parent,
     rest
   ) => {
-    return {
+    const captionNodes = node.children.filter(n => n.type !== 'emphasis')
+    const byline = node.children.find(n => n.type === 'emphasis')
+    const bylineNodes = byline.children || []
```
Comment (tpreusse) [inline components/editor/modules/figure/caption.js]: byline can be undefined

File: components/editor/modules/figure/caption.js [hunk 7]
```diff
@@ -35,16 +37,31 @@ const getSerializer = options => {
...
-        node.children,
-        0,
-        node,
-        rest
-      )
+      nodes: [
+        {
+          kind: 'block',
+          type: 'CAPTION_TEXT',
+          nodes: inlineSerializer.fromMdast(
```
Comment (tpreusse) [inline components/editor/modules/figure/caption.js]: please pass along all `fromMdast` arguments. Missing: index, parent and rest.

Same below

File: components/editor/modules/figure/caption.js (hunk 6 above)
Comment (tpreusse) [reply components/editor/modules/figure/caption.js]: and then explodes with `Cannot read property 'children' of undefined`
